### PR TITLE
Clean + comment MaxNLocator

### PR DESCRIPTION
**Pull request by dstansby**
- Remove redundant "always positive" comment (abs is always always positive!)
- Improve docstring for `steps` argument
- Clean up logic for filtering the step sizes in the classic autolimit mode. I thought using a loop for this before was quite opaque, so hope what I've done here is clearer?
- Add a comment above the loop through steps, explaining what it's doing.

### CODE REVIEW

File: lib/matplotlib/ticker.py [hunk 1]
```diff
@@ -1899,7 +1899,7 @@ def view_limits(self, dmin, dmax):
 
 
 def scale_range(vmin, vmax, n=1, threshold=100):
-    dv = abs(vmax - vmin)  # > 0 as nonsingular is called before.
```
Comment (anntzer) [inline lib/matplotlib/ticker.py:1902]: I think the point is that this is nonzero (>0, not >=0).

**Review by oscargus — APPROVED** [review]
Subject to re-adding the comment. It is quite important that dv > 0 since it is used as a denominator and log is computed on it. So while it will still be > 0, I think it is a worthwhile comment for later reference.

(I might consider micro-optimizing the code to change `if abs(meanv) / dv < threshold:` to `if abs(meanv) < threshold * dv:` though, but that is just me...

And possibly add a comment that `// 1` is floor. Had to think a bit there...)

File: lib/matplotlib/ticker.py (hunk 1 above)
Comment (dstansby) [reply lib/matplotlib/ticker.py:1902]: Ah makes sense - I'll put the comment back

File: lib/matplotlib/ticker.py [hunk 2]
```diff
@@ -2092,27 +2092,27 @@ def _raw_ticks(self, vmin, vmax):
...
         steps = self._extended_steps * scale
         if self._integer:
             # For steps > 1, keep only integer values.
             igood = (steps < 1) | (np.abs(steps - np.round(steps)) < 0.001)
             steps = steps[igood]
 
-        istep = np.nonzero(steps >= raw_step)[0][0]
-
-        # Classic round_numbers mode may require a larger step.
+        good_steps = steps >= ((_vmax - _vmin) / nbins)
```
Comment (anntzer) [inline lib/matplotlib/ticker.py:2101]: If I follow the code correctly (perhaps I'm not) at that point the steps will be [bad0, bad1, ..., badN, good0, good1, ..., goodN] but actually the steps that will be considered below will be (in that order) good0, badN, bad(N-1), ... and good1, ... will never be considered?  In which case I'd say the name "good_steps" is a bit confusing.

File: lib/matplotlib/ticker.py [hunk 3]
```diff
@@ -2092,27 +2092,27 @@ def _raw_ticks(self, vmin, vmax):
...
+            # Classic round_numbers mode may require a larger step.
+            # Get first multiple of steps that are <= _vmin
+            floored_vmins = (_vmin // steps) * steps
+            floored_vmaxs = floored_vmins + steps * nbins
+            large_steps = good_steps & (floored_vmaxs >= _vmax)
+
+        # Find index of smallest large step
+        istep = np.nonzero(large_steps)[0][0]
+
+        # Start at smallest good step, and check if this provides enough
```
Comment (anntzer) [inline lib/matplotlib/ticker.py:2112]: comment also needs updating

**Review by anntzer — COMMENTED** [review]
This seems OK, but the code is still actually quite confusing to me.  I am not actually sure whether the filtering of large_steps actually matters, or if it is just a small optimization for the `for step in steps...` loop below?  i.e. is it OK if we just loop through all steps (backwards) without prefiltering out large steps?
If not it may still be clearer to move the filtering into the loop, i.e.
```python
for step in steps[::-1]:
    if (the_condition_above_without_vectorization): continue # skip large steps
    do_the_usual_thing
```
Note that the steps array is small anyways so there isn't much to gain from the vectorization.

**Comment by dstansby** [comment]
I can't think of an easy way to remove the pre-filtering of `large_steps`. The core issue is that we have an array that looks like `[..., True, True, False, False...]`, and want to start at the index of the first `False` entry in the array. In order to do that I think there has to be some sort of pre-computation before the main loop to deterimine where this is.

I think as it stands this is a bit of an improvement - initially I was confused by `istep` being modified by the loop under `if mpl.rcParams['axes.autolimit_mode'] == 'round_numbers':`, and instead modifying the mask and then getting `istep` is a bit clearer. Very happy for other suggestions to improve more though, I definitely agree it's still not terribly clear.

**Review by anntzer — APPROVED** [review]

**Comment by anntzer** [comment]
Let's get this in.  Perhaps squash first?  Feel free to self-merge with or without squashing.

**Pull request merged**